server/app/app.py

```diff
@@ -1,26 +1,190 @@
 import os
 import datetime
 
-from flask import Flask
+from flask import Flask, redirect, current_app, make_response, render_template
+from flask import Blueprint, request, send_from_directory
 from flask_caching import Cache
 from flask_compress import Compress
 from flask_cors import CORS
+from flask_restful import Api, Resource
 
-from server.app.rest_api.rest import get_api_resources
-from server.app.util.utils import Float32JSONEncoder
-from server.app.web import webapp
+from http import HTTPStatus
+
+import server.common.rest as common_rest
+from server.common.utils import path_join, Float32JSONEncoder
+from server.data_common.matrix_loader import MatrixDataLoader, MatrixDataType
+
+from functools import wraps
+
+webbp = Blueprint("webapp", "server.common.web", template_folder="templates")
+
+
+@webbp.route("/")
+def index():
+    if current_app.app_config.datapath is None:
+        return dataroot_index()
+
+    dataset_title = current_app.app_config.title
+    scripts = current_app.app_config.scripts
+    return render_template("index.html", datasetTitle=dataset_title, SCRIPTS=scripts)
+
+
+@webbp.route("/favicon.png")
+def favicon():
+    return send_from_directory(os.path.join(webbp.root_path, "static/img/"), "favicon.png")
+
+
+def get_data_adaptor(dataset=None):
+    config = current_app.app_config
+
+    if dataset is None:
+        datapath = config.datapath
+    else:
+        datapath = path_join(config.dataroot, dataset)
+        # path_join returns a normalized path.  Therefore it is
+        # sufficient to check that the datapath starts with the
+        # dataroot to determine that the datapath is under the dataroot.
+        if not datapath.startswith(config.dataroot):
+            raise RuntimeError("Invalid dataset {dataset}")
```

Comment:
Given the very specific semantics indicated by this exception (ie, couldn't find the named dataset), we should be defining our own Exception class, and throwing that.  This prevents the catch block in `rest_get_data_adaptor` from accidentally catching some other "random" use of RuntimeError.

Reply:
fixed